OPT+BF: annotate read_only git/annex invocations, thread lock read-write git/annex and git-config invocations

yarikoptic opened the pull request:
old description which got it going:  While looking at py-spy stack from long running status I noticed that majority
of time is spent on git config call, triggered by checking for fake dates
config setting. AFAIK for read-only operations, which cannot produce a commit,
it might benefit to skip that check. And _call_git already has that kwarg,
but it is not proxied by "higher-level" helpers.  To avoid breeding workarounds
(like the one RFed within this commit) I decided that we might benefit from
exposing that option in higher-level helpers as well. Then "read-only" invocations could set it to False and we might avoid needless  config read in some cases (I have not done any timing on any prototypical use case which might benefit from this, but I would assume that status or diff might).

What is in the PR
- RF to make possible to make "lean" `call_git` read only invocations which would not even read config since not needed
- BF add thread locking non-read-only invocations of git/annex -- in `GitRepo.call_git` and `ConfigManager._run`.  That closes #5131 .  Now you can really animate your terminals with something like `datalad install -r -J50 /// ` ;-)

kyleam commented:
> While looking at py-spy stack from long running status I noticed that majority
> of time is spent on git config call, triggered by checking for fake dates
> config setting.

Hmm, I'd need to look into this again, but that doesn't sound right (i.e. if that's happening, I think it's a bug).  For a repo instance, the value should be cached at `_fake_dates_enabled`.

> AFAIK for read-only operations, which cannot produce a commit,
> it might benefit to skip that check. And _call_git already has that kwarg,
> but it is not proxied by "higher-level" helpers.

Note that this was explicitly discussed when the original series was proposed: https://github.com/datalad/datalad/pull/3791#discussion_r335355781

kyleam commented:
> For a repo instance, the value should be cached at `_fake_dates_enabled`.

That appears to be working as expected, at least for a simple case.

<details>
<summary>test</summary>

```diff
diff --git a/datalad/support/gitrepo.py b/datalad/support/gitrepo.py
index eab25a48e8..d5af0c3733 100644
--- a/datalad/support/gitrepo.py
+++ b/datalad/support/gitrepo.py
@@ -1603,6 +1603,7 @@ def fake_dates_enabled(self):
         """Is the repository configured to use fake dates?
         """
         if self._fake_dates_enabled is None:
+            print("--- config.getbool()")
             self._fake_dates_enabled = \
                 self.config.getbool('datalad', 'fake-dates', default=False)
         return self._fake_dates_enabled
```

```python
import tempfile
import datalad.api as dl

ds = dl.Dataset(tempfile.mkdtemp(prefix="dl-")).create()
ds.repo.call_git(["status"])
ds.repo.call_git(["status"])
```

```
[INFO   ] Creating a new annex repo at /tmp/dl-jwsbo0v0 
--- config.getbool()
[INFO   ] Scanning for unlocked files (this may take some time) 
```

</details>

yarikoptic commented:
> Hmm, I'd need to look into this again, but that doesn't sound right (i.e. if that's happening, I think it's a bug). For a repo instance, the value should be cached at `_fake_dates_enabled`.

yes, it is !!  But in my case it is LOTS (thousands) of small subdatasets.  So it does cache per dataset, but for a single dataset it doesn't need to do much ;) Actually, I was wrong about the status (or diff?): here is the flamegraph I got: http://www.onerussian.com/tmp/busy-datalad.svg  so it comes from `_save_add_submodules` in call to `get_hexsha` (read only, no need for fake date handling)
```
            subm_commit = subm.get_hexsha()
            if not subm_commit:
                yield get_status_dict(
                    action='add_submodule',
                    ds=self,
                    path=path,
                    status='error',
                    message=('cannot add subdataset %s with no commits', subm),
                    logger=lgr)
                continue
```
and then continues to the next one if not to be committed.  No need to consult config anywhere AFAIK, so no need to load all 4k of them if no submodule to be added.

yarikoptic commented:
hm, benchmarks reported
```
      7.75±0.04ms         9.30±1ms    ~1.20  support.path.get_parent_paths.time_allsubmods_toplevel
```
which is strange... I will restart to see if reproduces

yarikoptic commented:
took a longer py-spy trace: http://www.onerussian.com/tmp/busy-datalad-3.svg (for some reason no colors) -- different code path, just shows that: 1. caching of config worked and since it was first loaded, subsequent `fake_dates_enabled` does not trigger reload; 2. this PR will not be a "full solution" for that long running step since there is other long running steps (there is no logs at INFO or progressbar to tell them apart ATM).

mih commented:
Can we close this after #5076 was merged?

yarikoptic commented:
IMHO this one is largely orthogonal. Even though subdatasets use case was solved at higher level, and I don't have another affected use case, this pr changes make sense to me

kyleam commented:
Hmm, I'm on the fence about whether we should wait until having something concrete before deciding to expose this in each `call_git*` signature.  

If we do go with these changes, I think the commit message really should be updated to point to the previous discussion/decision on this and mention why it should be reversed (to account for the "many repos that wouldn't otherwise trigger a config read" usecase, IIUC).

yarikoptic commented:
yeap, adjusted commit message locally (not pushing for now) to

<details>
<summary>new commit msg</summary> 

    OPT: pass and use check_fake_dates kwargs through call call_git helpers
    
    While looking at py-spy stack from long running status I noticed that majority
    of time is spent on git config call, triggered by checking for fake dates
    config setting. AFAIK for read-only operations, which cannot produce a commit,
    it might benefit to skip that check. And _call_git already has that kwarg,
    but it is not proxied by "higher-level" helpers.  To avoid breeding workarounds
    (like the one RFed within this commit) I decided that we might benefit from
    exposing that option in higher-level helpers as well. Then "read-only" invocations
    could set it to False and we might avoid needless  config read in some cases
    (I have not done any timing on any prototypical use case which might benefit
    from this, but I would assume that status or diff might).
    
    We had an original discussion on either to expose check_fake_dates in higher
    level *Repo interfaces before:
    https://github.com/datalad/datalad/pull/3791#discussion_r335355781 and decided
    to not do it at that point.
    
    As for 'subdatasets' call, performance issue was now addressed with
    https://github.com/datalad/datalad/pull/5076 , so this change is not strictly
    necessary to optimize "subdatasets", but would still be generally benefitial
    for possible other invocations of *Repo methods across many instances without
    unnecessarily triggering loading of the config, typically in the case of
    read-only git operations (git-annex might need to merge git-annex branch, so
    typically annex calls should not disable fake dates).
</details>

but I wonder if any such RF to be done, should we may be introduce `read_only=False` instead? 
`check_fake_dates` is just current single (although nicely specific) operation, but it really boils down to either invoked `git` command is expected to be just "reading" or might "write" (anything which creates a commit -- `commit`, `merge`, etc) and thus then fake dates check should be invoked.

kyleam commented:
> I wonder if any such RF to be done, should we may be introduce `read_only=False` instead? [...]

Ah, good point.  I like that much better.

yarikoptic commented:
ok, redone it for `read_only=False` instead of `check_fake_dates=True` and then added locking on top for parallel execution in non-read_only mode in GitRepo and ConfigManager .  So it somewhat piled up two aspects - if desired can split the 2nd one in a follow up PR to keep our CIs warm.

yarikoptic commented:
heh, forgot that @mih just reported that crawler uses those older, now bad helpers,
```
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/datalad_crawler/nodes/annex.py", line 909, in _commit
    check_fake_dates=True)
  File "/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/datalad/support/gitrepo.py", line 329, in _wrap_normalize_paths
    result = func(self, files_new, *args, **kwargs)
TypeError: _git_custom_command() got an unexpected keyword argument 'check_fake_dates'
-------------------- >> begin captured logging << --------------------
asyncio: DEBUG: Using selector: EpollSelector
asyncio: DEBUG: Using selector: EpollSelector
```
I guess will need to fix up/release there, but cannot since then it would need first datalad released, unless we do go through a deprecation cycle here  with a kludge for check_fake_dates... eh heh... sure thing I could provide signature based kludge in the crawler -- not sure what would be better ;-)

yarikoptic commented:
for the crawler -- picking up on @mih's https://github.com/datalad/datalad-crawler/pull/82 but unfortunately it caused infamous stalls of batched processes, which we thought are gone, and for which we have mitigation (abandon) only in master ATM... might need to bring it in also into `maint`.

yarikoptic commented:
fixedup crawler released, restarted workflows.

kyleam commented on datalad/support/gitrepo.py:
```diff
@@ -2111,7 +2118,7 @@ def _git_custom_command(self, files, cmd_str,
                             shell=None, expect_fail=False,
-                            check_fake_dates=False,
+                            read_only=False,
                             index_file=None,
```
`_git_custom_command` is no longer used in the code base.  I think it'd be better to just leave it be.

yarikoptic replied:
I am ok either way (since likely to be removed if not used I guess).  Let's see if any other changes needed and I will adjust

yarikoptic replied:
ok, added reverting commit

kyleam commented:
With the lock changes, I've been unable to trigger the config issue from gh-5131.  As you say there, I also don't think using thread.Lock will add noticeable overhead.  So that part looks good to me.

kyleam commented on datalad/support/gitrepo.py:
```diff
@@ -2201,8 +2208,10 @@ def _call_git(self, args, files=None, expect_stderr=False, expect_fail=False,
-        check_fake_dates : bool, optional
-          If enabled (default), inspect if dates should be faked for dataset
+        read_only : bool, optional
+          If False (default), do not assume that command does not introduce changes and
```
I find the double negative hard to parse here.  When trying to come up with a rewrite, I stumbled on how to phrase it accurately.  I started to think about the larger issue with `read_only`: some commands in git are semantically read-only but actually modify the repo in ways that can clash with other running git commands.  For example, `status` and `diff` will refresh and write the index.  These writes are optimizations to avoid recomputing the same thing on the next git call.

In the case of `status`, the global `--no-optional-locks` option can be used to avoid these.  27344d6a6c (git: add --no-optional-locks option, 2017-09-27) makes it clear that this option doesn't cover (at least yet) all the commands that do such an "opportunistic" refresh.

So, while I think `read_only` (versus a more specific `check_fake_dates`) is appealing because it adds a single high-level knob that conceptually covers `check_fake_dates` (and now locking), my concern is that it's misleading because it doesn't reflect how git behaves underneath.

Anyway, perhaps not a reason to avoid adding `read_only`, but I think it's worth thinking about and making the docstring clearer.  Also, given the support for parallel operations now, our `status` calls in `.dirty` should use `--no-optional-locks`, and we should audit our `diff` calls to see if we can switch to a lower-level command (`diff-files`, ...) that doesn't refresh the index.  (But neither of those things need to be tied to this PR of course.)

As an unrelated note, this parameter should be documented in the other `call_git*` commands, perhaps with some fancy formatting like config's `_where_reload`, though I think nearby comments reminding future updaters to cover the other spots would do.

yarikoptic replied:
yeah, I was not 100% happy with read_only as well, but other ones I thought about are either not generic enough (e.g, `no_commit`) or vague (e.g. `changes_state`).
Please suggest changes to the wording - I did think about making it also more specific, like "no commits etc", but just makes it longer for no good value.  We could add a disclaimer/note that "Note: even if 'read-only' commands could refresh index etc." may be, but again -- not 100% certain if such level of detail is really needed here

yarikoptic replied:
adjusted description a bit, hopefully it is better

kyleam replied:
> I did think about making it also more specific, like "no commits etc", but just makes it longer for no good value.

Or it actually makes it incorrect.  The problem isn't that the parameter name is inaccurate; it's that the caller may think some commands are clearly read-only, but in fact they write to the repository.  In those cases, it wouldn't be appropriate to skip the locks needed for parallel operations.

kyleam commented on datalad/support/gitrepo.py:
```diff
@@ -2201,10 +2208,12 @@ def _call_git(self, args, files=None, expect_stderr=False, expect_fail=False,
-          commits and set up the Git environment to ensure that. Must be
-          disabled for repository initialization.
+        read_only : bool, optional
+          If set to True, we assume that command does not introduce changes to recorded
+          in git information (commits, heads, tags) and merely reads/outputs information
+          (note: git might still adjust caches and index).
+          If set to False, we would inspect if dates should be faked for dataset
+          commits and set up the Git environment to ensure that.
```
How about something like this?  I'm trying to make it clearer that `read_only=True` isn't appropriate for the cases where git does an opportunistic refresh.

```suggestion
          By setting this to True, the caller indicates that the command does
          not write to the repository, which lets this function skip some
          operations that are necessary only for commands the modify the
          repository. Beware that even commands that are conceptually
          read-only, such as `git-status` and `git-diff`, may refresh and write
          the index.
```

kyleam commented on datalad/support/annexrepo.py:
```diff
@@ -1620,7 +1621,8 @@ def _check_files(self, fn, files, batch):
         modified = [
             f for f in self.call_git_items_(
-                ['diff', '--name-only', '-z'], sep='\0')
+                ['diff', '--name-only', '-z'], sep='\0',
```
As mentioned in my other response, I think `git-diff` is an example of a command that may refresh the index and is not safe to mark as `read_only=True`, provided that we're using `read_only` to decide whether to take a lock for parallel operations.


```suggestion
                ['diff-files', '--name-only', '-z'], sep='\0',
```

yarikoptic commented:
Thank you @kyleam - applied adjusted wording (i like it), and switch from `diff`

kyleam reviewed (approved):
Thanks.  Looks good to me.

I think all of the `call_git*` docstrings should reference the parameter, perhaps just with `call_git` holding the actual text and the description for all of the other methods pointing to `call_git`s description.  But I can propose it in a follow-up PR (and either way it's not a big deal).

yarikoptic commented:
Restarted failed osx and appveyor - don't think the failures are related to pr but who knows

yarikoptic commented:
<details>
<summary>OSX failures persist so must be related since otherwise I think OSX runs are green.  grr...  I will restart once more (only OSX) and will try to look into it - this time recording a sample failed test for comparison</summary> 

```shell
======================================================================
FAIL: datalad.support.tests.test_annexrepo.test_AnnexRepo_file_has_content(False,)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/Users/runner/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/site-packages/datalad/tests/utils.py", line 1090, in _wrap_with_testrepos
    t(*(arg + (uri,)), **kw)
  File "/Users/runner/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/site-packages/datalad/tests/utils.py", line 755, in _wrap_with_tempfile
    return t(*(arg + (filename,)), **kw)
  File "/Users/runner/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/site-packages/datalad/support/tests/test_annexrepo.py", line 319, in test_AnnexRepo_file_has_content
    [ar.supports_unlocked_pointers])
AssertionError: [False] != [True]
-------------------- >> begin captured logging << --------------------
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
    return t(*(arg + (filename,)), **kw)
  File "/Users/runner/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/site-packages/datalad/support/tests/test_annexrepo.py", line 354, in test_AnnexRepo_is_under_annex
    [ar.supports_unlocked_pointers])
AssertionError: [False] != [True]
-------------------- >> begin captured logging << --------------------
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
asyncio: DEBUG: Using selector: KqueueSelector
```
</details>

yarikoptic commented:
FWIW, the same failures happen for me on OSX on our test OSX box when I run the same full selection of tests on master, but they do not happen if I run just a single test (e.g. `python -m nose -s -v --with-cov --cover-package datalad --cover-xml ../datalad/support/tests/test_annexrepo.py:test_AnnexRepo_file_has_content `).

yarikoptic commented:
it is "fun": apparently the important difference in either specifying by path or by module path, i.e. `python -m nose -s -v --with-cov --cover-package datalad --cover-xml datalad.support.tests.test_annexrepo:test_AnnexRepo_file_has_content` would fail whenever `python -m nose -s -v --with-cov --cover-package datalad --cover-xml ../datalad/support/tests/test_annexrepo.py:test_AnnexRepo_file_has_content` would be ok

yarikoptic commented:
grrr, f.cking bug ran away somehow.

<details>
<summary>here is some incomplete (forgot to copy the last line but I 99% certain it was True ;)) copy/paste I managed to catch into my clipboard history</summary> 

```shell
> /Users/yoh/proj/datalad/datalad/venvs/dev3/lib/python3.8/site-packages/datalad/support/tests/test_annexrepo.py(318)test_AnnexRepo_file_has_content()
-> eq_(ar.file_has_content(["test-annex.dat"], batch=batch),
(Pdb) p ar.file_has_content(["test-annex.dat"], batch=batch)
[False]
(Pdb) p ar.file_has_content(["test-annex.dat"], batch=batch)
[False]
(Pdb) p ar.file_has_content(["test-annex.dat"], batch=batch)
[False]
(Pdb) p batch
False
(Pdb) p ar.supports_unlocked_pointers
True
(Pdb) p ar.file_has_content(["test-annex.dat"], batch=batch)
[False]
(Pdb) p ar.file_has_content(["test-annex.dat"], batch=batch)
[False]
(Pdb) p ar
AnnexRepo('/private/var/folders/wr/944pwmkx3yg0r2sf1g3zpkrh0000gn/T/datalad_temp_test_AnnexRepo_file_has_content30u1oh06')
(Pdb) 
[1]+  Stopped                 DATALAD_TESTS_TEMP_KEEP=1 python -m nose -s -v --with-cov --cover-package datalad --pdb --pdb-failures --cover-xml datalad.support.tests.test_annexrepo:test_AnnexRepo_file_has_content
content30u1oh06evel) bash-3.2$ cd /private/var/folders/wr/944pwmkx3yg0r2sf1g3zpkrh0000gn/T/datalad_temp_test_AnnexRepo_file_has_ 
(dev3) (pyout-devel) bash-3.2$ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        typechange: test-annex.dat

(dev3) (pyout-devel) bash-3.2$ fg
DATALAD_TESTS_TEMP_KEEP=1 python -m nose -s -v --with-cov --cover-package datalad --pdb --pdb-failures --cover-xml datalad.support.tests.test_annexrepo:test_AnnexRepo_file_has_content (wd: ~/proj/datalad/datalad/__testhome__)
^[[A
p ar
AnnexRepo('/private/var/folders/wr/944pwmkx3yg0r2sf1g3zpkrh0000gn/T/datalad_temp_test_AnnexRepo_file_has_content30u1oh06')
(Pdb) p ar.file_has_content(["test-annex.dat"], batch=batch)

```
</details>

which suggests that somehow `git status` made things straight, i.e. so that `file_has_content(` started to report correct `True` after I ran `git status` outside.
My hypothesis is that with this OPT we managed to avoid some `git config` call which had similar "positive" effect and avoided some kind of a race condition somewhere... but not yet sure where/how 

<details>
<summary>note: appveyor failure is also about file_has_content - just in a different test</summary> 

```shell
======================================================================
FAIL: datalad.support.tests.test_annexrepo.test_AnnexRepo_web_remote
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Miniconda37\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "C:\Miniconda37\lib\site-packages\datalad\tests\utils.py", line 576, in _wrap_with_tree
    return t(*(arg + (d,)), **kw)
  File "C:\Miniconda37\lib\site-packages\datalad\tests\utils.py", line 688, in _wrap_serve_path_via_http
    return tfunc(*(args + (path, url)), **kwargs)
  File "C:\Miniconda37\lib\site-packages\datalad\tests\utils.py", line 755, in _wrap_with_tempfile
    return t(*(arg + (filename,)), **kw)
  File "C:\Miniconda37\lib\site-packages\datalad\support\tests\test_annexrepo.py", line 384, in test_AnnexRepo_web_remote
    ok_(ar.file_has_content(testfile))
AssertionError: None
-------------------- >> begin captured logging << --------------------
asyncio: DEBUG: Using proactor: IocpProactor
```
</details>
may be it could be the one to track down instead...

yarikoptic commented:
on windows it boils down to the switch from `diff` to `diff-files` and here is the difference in the logged output
```
-datalad.cmd: DEBUG: Async run ['git', 'diff', '--name-only', '-z']^M
-datalad.cmd: DEBUG: Launching process ['git', 'diff', '--name-only', '-z']^M
-datalad.cmd: DEBUG: Process 19716 started^M
-datalad.cmd: DEBUG: Waiting for process 19716 to complete^M
-datalad.cmd: Level 5: Read 132 bytes from 19716[stderr]:^M
-datalad.cmd: Level 5: stderr| warning: LF will be replaced by CRLF in 127.0.0.1_about.txt.^M
-| The file will have its original line endings in your working directory^M
-|  ^M
-datalad.cmd: DEBUG: Process 19716 exited with return code 0^M
-datalad.cmd: Level 8: Finished running ['git', 'diff', '--name-only', '-z'] with status 0^M
-datalad.gitrepo: Level 11: stderr| warning: LF will be replaced by CRLF in 127.0.0.1_about.txt.^M
-datalad.gitrepo: Level 11: stderr| The file will have its original line endings in your working directory^M
+datalad.cmd: DEBUG: Async run ['git', 'diff-files', '--name-only', '-z']^M
+datalad.cmd: DEBUG: Launching process ['git', 'diff-files', '--name-only', '-z']^M
+datalad.cmd: DEBUG: Process 10040 started^M
+datalad.cmd: DEBUG: Waiting for process 10040 to complete^M
+datalad.cmd: Level 5: Read 20 bytes from 10040[stdout]:^M
+datalad.cmd: Level 5: stdout| 127.0.0.1_about.txt^@ ^M
+datalad.cmd: DEBUG: Process 10040 exited with return code 0^M
+datalad.cmd: Level 8: Finished running ['git', 'diff-files', '--name-only', '-z'] with status 0^M
``` 
and what confuses me is that int he case of `diff` (the one where test passes) there is no stdout logged at all, just all the warnings in the stderr (none of that in diff-files). any immediate clues @kyleam ?

kyleam commented:
> on windows it boils down to the switch from `diff` to `diff-files` and here is the difference in the logged output [...] any immediate clues @kyleam ?

No, but I can trigger it with `tools/eval_under_testloopfs`, so something with adjusted branches (maybe to due with filters).  I'll dig more.

kyleam commented:
Under `tools/eval_under_testloopfs`, here's the repo from `test_AnnexRepo_web_remote` immediately before the first `file_has_content` call:

```console
$ git diff-files -p
$ git diff-files
:100755 100755 6712b5d7d14789b99e55540b0dec07fe2fc0d421 0000000000000000000000000000000000000000 M	127.0.0.1_about.txt
$ git diff
$ git diff-files
```

So, we're switching to `diff-files` because it doesn't do a refresh of the index, but I suppose that that comes along that comes along with not triggering the clean filter.

In this case, I'd say it makes sense to switch this spot back `"diff", read_only=False`.

yarikoptic commented:
eh, ok, will do... since anything else (checking if we are in adjusted branch etc) would only take more time, thus ruining any OPT effect.  I would expect OSX also to turn green, but I still think it would only be hiding the issue - I think there is more to it there.

kyleam commented:
> since anything else (checking if we are in adjusted branch etc) would only take more time, thus ruining any OPT effect.

I don't think any of this is for this PR but...

Another option would be to switch to `ls-files --modified PATH`, which seems to correctly run the filter yet doesn't opportunistically refresh the index.  I don't think that's a step worth taking though, because in my view `file_has_content` should just be rewritten to use a combination of `diffstatus` (which calls `ls-files --modified`) and `get_content_annexinfo`.  Also, it seems like we should consider whether `get_content_annexinfo` ought to report that content is present if a file is modified, as it does now; if we decide it shouldn't, then we don't really need `file_has_content` and it can be reduced to a very thin (perhaps deprecated) wrapper around `get_content_annexinfo`.

And I think it's worth keeping in mind that avoiding the thread lock by disabling the opportunistic refresh of the index might not be a net gain because downstream the index update would need to be redone.  Perhaps tricky, but it'd bee good to measure in some way rather than assuming there's actually a benefit.

Unrelated to any of that: the fact that `ls-files --modified` and `diff-files` when combined with flags like `-p` and `-s` do seem to run the filter makes me think that `diff-files` not running the filter in some cases should be looked into more and reported upstream.

> I would expect OSX also to turn green, but I still think it would only be hiding the issue - I think there is more to it there.

I'm confused by this.  Based on you saying "OSX tests were failing even at 4d84e6e so before this switch to diff-files was introduced", I thought those were unrelated.

yarikoptic commented:
> > I would expect OSX also to turn green, but I still think it would only be hiding the issue - I think there is more to it there.
> 
> I'm confused by this. Based on you saying "OSX tests were failing even at [4d84e6e](https://github.com/datalad/datalad/commit/4d84e6ec063706a6c35e47cebf4b569d155c3e0c) so before this switch to diff-files was introduced", I thought those were unrelated.

but they were `read_only=True`.  My hypothesis [was](https://github.com/datalad/datalad/pull/5070#issuecomment-736907282) that some `git config` call we do in case of non-read_only does something "magically necessary" to avoid those failures.  And BTW indeed those OSX builds turned green!  Travis is still cranking, and crippledFS failures were closed eyes upon for awhile so would be too much effort to say if related ATM.

yarikoptic commented:
ok, let's proceed!
